clang/lib/CConv/AVarBoundsInfo.cpp

```diff
@@ -1071,36 +1201,49 @@ bool AVarBoundsInfo::performFlowAnalysis(ProgramInfo *PI) {
       removeBounds(TBK, FlowInferred);
   }
 
-  // Next, get the ARR pointers that has bounds.
-  // These are pointers with bounds.
-  std::set<BoundsKey> ArrWithBounds;
-  for (auto &T : BInfo) {
-    ArrWithBounds.insert(T.first);
-  }
-  // Also add arrays with invalid bounds.
-  ArrWithBounds.insert(InvalidBounds.begin(), InvalidBounds.end());
+  std::set<BoundsKey> ArrNeededBounds, ArrNeededBoundsNew;
+  ArrNeededBounds.clear();
 
-  // This are the array atoms that need bounds.
-  // i.e., ArrNeededBounds = ArrPtrs - ArrPtrsWithBounds.
-  std::set<BoundsKey> ArrNeededBounds;
-  std::set_difference(ArrPointers.begin(), ArrPointers.end(),
-                      ArrWithBounds.begin(), ArrWithBounds.end(),
-                      std::inserter(ArrNeededBounds, ArrNeededBounds.end()));
+
+  getBoundsNeededArrPointers(ArrPointers, ArrNeededBounds);
 
   bool Changed = !ArrNeededBounds.empty();
 
   // Now compute the bounds information of all the ARR pointers that need it.
+  // We iterate until there are no new array variables whose bounds are found.
+  // The expectation is every iteration we will find bounds for at least one array
+  // variable.
   while (Changed) {
-    Changed = false;
-    // Use flow-graph.
-    Changed = performWorkListInference(ArrNeededBounds, this->ProgVarGraph) || Changed;
-    // Use potential length variables.
-    Changed = performWorkListInference(ArrNeededBounds, this->ProgVarGraph, true) || Changed;
-    // Try solving using context-sensitive BoundsKey.
-    Changed = performWorkListInference(ArrNeededBounds, this->CtxSensProgVarGraph) || Changed;
-    Changed = performWorkListInference(ArrNeededBounds, this->CtxSensProgVarGraph, true) || Changed;
+    // Clear all inferred bounds.
+    ABI.clearInferredBounds();
+    // Regular flow inference.
+    performWorkListInference(ArrNeededBounds, this->ProgVarGraph, ABI);
+    // Flow inference using potential bounds.
+    performWorkListInference(ArrNeededBounds, this->ProgVarGraph, ABI, true);
```

Comment:
What is a "potential bound" (`fromPB`) ?

Reply:
As mentioned in the pull request:

For other variables that do not have any choice of bounds, we use potential bounds choices (`FromPB`), these are the variables that are upper bounds to an index variable used in an array indexing operation.
For example:
```
if (i < n) {
   ...arr[i]...
}
```
In the above case, we use `n` as a potential count bounds for arr.
__Note: we only use potential bounds for a variable when none of its predecessors have bounds.__